src/eos-updater-util.c

```diff
@@ -160,127 +153,136 @@ static const gchar *dmi_attributes[] =
     NULL,
   };
 
-gboolean
-eos_updater_resolve_upgrade (EosUpdater  *updater,
-                             OstreeRepo *repo,
-                             gchar     **upgrade_refspec,
-                             gchar     **original_refspec,
-                             gchar     **booted_checksum,
-                             GError    **error)
+static OstreeDeployment *
+get_booted_deployment (GError **error)
 {
-  gboolean ret = FALSE;
-  gboolean on_hold = FALSE;
-  guint status = 0;
-  gs_free gchar *o_refspec = NULL;
-  gs_free gchar *o_remote = NULL;
-  gs_free gchar *o_ref = NULL;
-  gs_free gchar *vendor = NULL;
-  gs_free gchar *product = NULL;
-  gs_free gchar *p_id = NULL;
-  gs_free gchar *p_ref = NULL;
-  gs_free gchar *osgroup = NULL;
-  gs_free gchar *baseurl = NULL;
-  gs_free gchar *uri = NULL;
-  gs_free gchar *query = NULL;
-  gs_unref_object OstreeDeployment *merge_deployment = NULL;
-  const gchar *osname;
-  const gchar *booted;
-  GKeyFile *origin;
-  GKeyFile *repo_config;
-  GHashTable *hw_descriptors = NULL;
-  gs_unref_object SoupSession *soup = NULL;
-  gs_unref_object SoupMessage *msg = NULL;
-  gs_unref_keyfile GKeyFile *bkf = NULL;
-
-  OstreeSysroot *sysroot = ostree_sysroot_new_default ();
+  gs_unref_object OstreeSysroot *sysroot = ostree_sysroot_new_default ();
+  OstreeDeployment *booted_deployment = NULL;
 
   if (!ostree_sysroot_load (sysroot, NULL, error))
-    goto out;
+    return NULL;
 
-  if (!ostree_sysroot_get_booted_deployment (sysroot))
+  booted_deployment = ostree_sysroot_get_booted_deployment (sysroot);
+  if (booted_deployment == NULL)
     {
       g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
                            "Not an ostree system");
-      goto out;
+      return NULL;
     }
 
-  merge_deployment = ostree_sysroot_get_merge_deployment (sysroot, NULL);
-  osname = ostree_deployment_get_osname (merge_deployment);
-  origin = ostree_deployment_get_origin (merge_deployment);
-  booted = ostree_deployment_get_csum (merge_deployment);
+  return g_object_ref (booted_deployment);
```

Comment:
`ostree_sysroot_get_booted_deployment` does not transfer the ownership, so if we don't ref it, the booted deployment might be deleted when we unref the sysroot. It used to work without adding a ref because sysroot was actually leaking its deployments earlier (https://github.com/ostreedev/ostree/pull/287).